lib/chained_job/process.rb

```diff
@@ -72,5 +81,9 @@ def redis_key
     def job_key
       Helpers.job_key(job_arguments_key)
     end
+
+    def push_job_arguments_back
+      ChainedJob.redis.rpush(redis_key, Helpers.serialize([argument]))
```

Comment:
why is it wrapped into array?

Reply:
Yes, it's a major thing to fully understand by myself. I'm digging it down. Currently, this pull request is written as an idea. I will wrote down tests and find why we need to put in []. So keep it with me and I will find why.

Reply:
@raimondasv that's actually because of this:

Worker method ```array_of_job_arguments``` always returns an array [].
While we set ChainedJob::StoreJobArguments.run with .each_slice
```
array_of_job_arguments.each_slice(config.arguments_batch_size) do |sublist|
  ChainedJob.redis.rpush(redis_key, Helpers.serialize(sublist))
end
```

Reply:
And in general ... hm it would be better to reuse the same call instead of pushing myself. But needs refactor.

Reply:
I'm not sure about this. `lpop` operates on one argument without wrapped arrays, but here we push it back with extra wrapped array.

Reply:
I'ts redis stuff, rpush accepts element or elements, so array will be unpacked anyway:

```[15] pry(main)> MD.redis.rpush('hello', [1])
=> 1
[16] pry(main)> MD.redis.lpop('hello')
=> "1"

[7] pry(main)> MD.redis.rpush('hello', [1, 2, 3])
=> 3
[8] pry(main)> MD.redis.lpop('hello')
=> "1"
[9] pry(main)> MD.redis.lpop('hello')
=> "2"
[10] pry(main)> MD.redis.lpop('hello')
=> "3"


[4] pry(main)> MD.redis.rpush('hello', 1)
=> 1
[5] pry(main)> MD.redis.lpop('hello')
=> "1"
```

Reply:
@tomasv I did some debug and maybe this gets more clear:

```
arguments = [1, 2, 3]

ChainedJob.redis.rpush(redis_key, Helpers.serialize(sublist))
 ->
  def serialize(arguments)
    arguments.map { |argument| Marshal.dump(argument) }
  end -> ["\x04\bi\x06", "\x04\bi\a", "\x04\bi\b"]

->  redis.rpush(.., ["\x04\bi\x06", "\x04\bi\a", "\x04\bi\b"])
  -> redis.lrange('ax', 0, -1)
    => ["\u0004\bi\u0006", "\u0004\bi\a", "\u0004\bi\b"]

#### During the process:

ChainedJob.redis.lpop(redis_key)
 -> "\u0004\bi\u0006"
   -> Marshal.load(\u0004\bi\u0006")
    -> 1

During failure -> 1

Helpers.serialize([1])
 -> [1].map { |argument| Marshal.dump(argument) }
   -> ["\x04\bi\x06"]
     -> redis.rpush(.., ["\x04\bi\x06"])

-> redis.lrange('ax', 0, -1)
 => ["\u0004\bi\a", "\u0004\bi\b", "\u0004\bi\u0006"]

["\u0004\bi\u0006", "\u0004\bi\a", "\u0004\bi\b"]
["\u0004\bi\a", "\u0004\bi\b", "\u0004\bi\u0006"]

Both lines are the same but in a different order, as expected.
```

Reply:
@tomasv let me know if all good and I can merge and tag new version